lib/cdo/delete_accounts_helper.rb

```diff
@@ -0,0 +1,184 @@
+  # @param [User] user to be anonymized.
+  def self.anonymize_user(user)
+    user.clear_user_and_mark_purged
+
+    UserGeo.where(user_id: user.id).each(&:clear_user_geo)
+
+    SignIn.where(user_id: user.id).destroy_all
+  end
+
+  # Cleans all sections owned by the user.
+  # @param [Integer] The ID of the user to anonymize the sections of.
+  def self.anonymize_user_sections(user_id)
+    Sections.with_deleted.where(user_id: user_id).each(&:clean_data)
+  end
+
+  # Removes all information about the user pertaining to Pardot. This encompasses Pardot itself, the
+  # contact_rollups pegasus table (master and reporting), and the contact_rollups_daily pegasus table
+  # (reporting only).
+  # @param [Integer] The user ID to purge from Pardot.
+  def self.remove_from_pardot(user_id)
+    pardot_ids = PEGASUS_DB[:contact_rollups].
+      select(:pardot_id).
+      where(dashboard_user_id: user_id).
+      map {|contact_rollup| contact_rollup[:pardot_id]}
+    failed_ids = Pardot.delete_prospects(pardot_ids)
+    raise "Pardot.delete_prospects failed for Pardot IDs #{failed_ids.join(', ')}." unless failed_ids
+
+    PEGASUS_DB[:contact_rollups].where(dashboard_user_id: user_id).delete
+
+    PEGASUS_REPORTING_DB[:contact_rollups_daily].where(dashboard_user_id: user_id).delete
+  end
+
+  # Removes the SOLR record associated with the user.
+  # WARNING: This does not remove SOLR records associated with forms for the user.
+  # @param [Integer] The user ID to purge from SOLR.
+  def self.remove_from_solr(user_id)
+    SolrHelper.delete_document(SOLR, 'user', user_id)
+  end
+
+  # Purges (deletes and cleans) various pieces of information owned by the user in our system.
+  # Noops if the user is already marked as purged.
+  # @param [User] user The user to purge.
+  def self.purge_user(user)
+    # This early exit is necessary, else we would enter an infinite cycle if two users were both
+    # students of the other (via `purge_orphaned_students`).
+    return if user.purged_at
+
+    # It is important that user.destroy happen first, as `purge_orphaned_students` assumes the
+    # dependent destroys have already been executed.
+    user.destroy
+    # Note that we do not gate any deletion logic on `user.user_type` as its current state may not
+    # be reflective of past state.
+    DeleteAccountsHelper.clean_level_source_backed_progress(user.id)
+    DeleteAccountsHelper.clean_survey_responses(user.id)
+    DeleteAccountsHelper.delete_project_backed_progress(user.id)
+    DeleteAccountsHelper.purge_orphaned_students(user.id)
+    DeleteAccountsHelper.clean_and_destroy_pd_content(user.id)
+    DeleteAccountsHelper.anonymize_user(user)
+    DeleteAccountsHelper.anonymize_user_sections(user.id)
+    DeleteAccountsHelper.remove_from_pardot(user.id)
+    DeleteAccountsHelper.remove_from_solr(user.id)
+
+    user.purged_at = Time.zone.now
+    user.save(validate: false)
```

Comment:
No, purged users are expected to pass validation, I've gone to some trouble to make sure of that. That said, I want the `purged_at` update to get persisted, whether or not the user is valid.